scheduler/preemption.go

```diff
@@ -0,0 +1,572 @@
+// per job/task group. This is used while scoring preemption options
+func (p *Preemptor) SetPreemptions(allocs []*structs.Allocation) {
+	// Clear out existing values since this can be called more than once
+	for k := range p.currentPreemptions {
+		delete(p.currentPreemptions, k)
+	}
+
+	// Initialize counts
+	for _, alloc := range allocs {
+		id := structs.NamespacedID{alloc.JobID, alloc.Namespace}
+		countMap, ok := p.currentPreemptions[id]
+		if !ok {
+			countMap = make(map[string]int)
+		}
+		c := countMap[alloc.TaskGroup]
+		countMap[alloc.TaskGroup] = c + 1
+		p.currentPreemptions[id] = countMap
+	}
+}
+
+// getNumPreemptions counts the number of other allocations being preempted that match the job and task group of
+// the alloc under consideration. This is used as a scoring factor to minimize too many allocs of the same job being preempted at once
+func (p *Preemptor) getNumPreemptions(alloc *structs.Allocation) int {
+	numCurrentPreemptionsForJob := 0
+	countMap := p.currentPreemptions[structs.NamespacedID{alloc.JobID, alloc.Namespace}]
+	if countMap != nil {
+		numCurrentPreemptionsForJob = countMap[alloc.TaskGroup]
+	}
+	return numCurrentPreemptionsForJob
+}
+
+// PreemptForTaskGroup computes a list of allocations to preempt to accommodate
+// the resources asked for. Only allocs with a job priority < 10 of jobPriority are considered
+// This method is meant only for finding preemptible allocations based on CPU/Memory/Disk
+func (p *Preemptor) PreemptForTaskGroup(resourceAsk *structs.AllocatedResources) []*structs.Allocation {
+	resourcesNeeded := resourceAsk.Comparable()
+
+	// Subtract current allocations
+	for _, alloc := range p.currentAllocs {
+		p.nodeRemainingResources.Subtract(alloc.ComparableResources())
+	}
+
+	// Group candidates by priority, filter out ineligible allocs
+	allocsByPriority := filterAndGroupPreemptibleAllocs(p.jobPriority, p.currentAllocs)
+
+	var bestAllocs []*structs.Allocation
+	allRequirementsMet := false
+	var preemptedResources *structs.ComparableResources
+
+	// Iterate over allocations grouped by priority to find preemptible allocations
+	for _, allocGrp := range allocsByPriority {
+		for len(allocGrp.allocs) > 0 && !allRequirementsMet {
+			closestAllocIndex := -1
+			bestDistance := math.MaxFloat64
+			// Find the alloc with the closest distance
+			for index, alloc := range allocGrp.allocs {
+				currentPreemptionCount := p.getNumPreemptions(alloc)
+				maxParallel := 0
+				tg := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
+				if tg != nil && tg.Migrate != nil {
+					maxParallel = tg.Migrate.MaxParallel
+				}
+				distance := scoreForTaskGroup(resourcesNeeded, alloc.ComparableResources(), maxParallel, currentPreemptionCount)
+				if distance < bestDistance {
+					bestDistance = distance
+					closestAllocIndex = index
+				}
+			}
+			closestAlloc := allocGrp.allocs[closestAllocIndex]
+
+			if preemptedResources == nil {
+				preemptedResources = closestAlloc.ComparableResources()
```

Comment:
You don't need both preemptedResources and availableResources. Just initialized availableResources to nodeRemainingResources and then add the closest to it as you mark them for preemption